Add Create tab into Digital Twins page preview

VanessaScherma opened the pull request:
## Title

Add Create tab into Digital Twins page preview

## Type of Change

- [x] New feature
- [ ] Bug fix
- [ ] Documentation update
- [ ] Refactoring
- [ ] Security patch
- [ ] UI/UX improvement

## Description

PR with the aim of solving issue #1002.

When opening, the Create page shows the Editor and three files already placed in the sidebar:
- `description.md`
- `README.md`
- `.gitlab-ci.yml`

The user can:
- insert a new file via the ‘Add new file’ button (it is not possible to insert files with the same name, but the extension can be different)
- delete a file (either inserted by him or one of the predefined ones, but it is not possible to delete `.gitlab-ci.yml` as it is necessary for the execution of the DT)
- rename a file inserted by him

By clicking on the ‘Create’ button, the user can enter the name of the DT, which will correspond to the name of the sub-folder within his GitLab digitaltwins folder, and confirm the operation.
At the end of the operation, the Snackbar informs of success or failure.

By switching to one of the other tabs, the DT is already visible.

## Testing

Unit and integration tests are currently in progress.

## Impact

These changes add a new functionality, allowing the user to insert new digital twins.

## Additional Information

None.

## Checklist

- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have added tests for all the new code and any changes made to
      existing code.
- [ ] I have made corresponding changes to the documentation.

prasadtalasila commented:
@VanessaScherma Thanks for the PR. A very top-level review.

### What works:

1. Saving of skeletal DT in the gitlab repository

### What does not work:

1. The newly created DT does not show up in Manage and Execute pages
2. The new DT is not shown even after restarting the client and also not after relogin

### What are the suggested improvements:

1. Have a field for name of DT
2. Permit users to save DTs even if the files are empty

VanessaScherma commented:
@prasadtalasila 

The DT created should be immediately visible within the page, as shown in these screenshots:
![CreateDT](https://github.com/user-attachments/assets/bd90644b-2b90-479c-8cac-f2fb999a24cd)
![Snackbar](https://github.com/user-attachments/assets/bfbf02ef-c35c-4c5e-817d-e49a7f42bcca)
![Manage](https://github.com/user-attachments/assets/482c8d91-0947-4dc5-9747-2c18a675d58d)
In the example shown, I created a digital twin called ‘testDT’, correctly created as informed by the Snackbar at the bottom of the page after confirming the save. As soon as I changed tabs to Manage, it was immediately visible.

Regarding suggestions, I can move the field for the DT name from the Dialog to above the Editor, so that it is always visible. For empty files, on the other hand, I had initially left this option but a problem had arisen with Editor Monaco. In fact, if the file is empty, the text shown on EditorTab cannot be updated, showing the contents of the previously selected file. This is probably due to the ‘read-only’ option I added to avoid writing without having selected a file first. I can look for some alternative solution if you think the possibility of creating empty files is more suitable.

prasadtalasila commented:
@VanessaScherma 

> In the example shown, I created a digital twin called ‘testDT’, correctly created as informed by the Snackbar at the bottom of the page after confirming the save. As soon as I changed tabs to Manage, it was immediately visible.

Strange. I've created two [new DTs](https://maestro.cps.digit.au.dk/gitlab/dtaas/TestUserDTaaS/-/tree/main/digital_twins?ref_type=heads), namely "new-dt" and "new-digital-twin". But they don't show up in Manage and Execute tabs.

> Regarding suggestions, I can move the field for the DT name from the Dialog to above the Editor, so that it is always visible.

Ok thanks

> For empty files, on the other hand, I had initially left this option but a problem had arisen with Editor Monaco. In fact, if the file is empty, the text shown on EditorTab cannot be updated, showing the contents of the previously selected file. This is probably due to the ‘read-only’ option I added to avoid writing without having selected a file first. I can look for some alternative solution if you think the possibility of creating empty files is more suitable.

It is alright for now. Please focus on completing the PR.

prasadtalasila commented on client/src/preview/route/digitaltwins/DigitalTwinsPreview.tsx:
```diff
@@ -1,96 +1,76 @@
-      if (assets) {
-        createDigitalTwinsForAssets(assets, dispatch);
-      }
+    const defaultFiles = [
```
This array and the method to iterate upon this can moved to either to `gitlabDigitalTwin.ts`

VanessaScherma replied:
It has been moved to `file.ts` in `preview/util`.

prasadtalasila commented on client/src/preview/route/digitaltwins/create/CreateDTDialog.tsx:
The logic in this file is too low-level. I think most of these functions belong to either `util/digitalTwinUtils.ts` or `util/gitlabDigitalTwin.ts`.

VanessaScherma replied:
The functions have been moved to `init.ts` and `file.ts` in `preview/util`.

prasadtalasila commented on client/src/preview/util/gitlabDigitalTwin.ts:
please rename this file to `digitalTwin.ts`

prasadtalasila commented on client/src/preview/route/digitaltwins/create/CreateDialogs.tsx:
This is a better structure over `CreateDTDialog.tsx`

prasadtalasila commented on client/src/preview/route/digitaltwins/create/CreatePage.tsx:
```diff
@@ -0,0 +1,182 @@
+  setNewDigitalTwinName: Dispatch<SetStateAction<string>>;
+}
+
+function FileActionButtons({
```
please move this function to a new `File.ts`

prasadtalasila commented on client/src/preview/route/digitaltwins/editor/sidebarFunctions.tsx:
The broader problem I see is this.
We need the user to forget the existence of gitlab but understand DT abstraction. In addition, user knows (rightly) that a DT consists of multiple files including configuration file(s). Thus we need to have a dedicated code for modifying files which can be accessed by DT class and if needed gitlab class. The code dependency graph I see is:

```txt
DigitalTwin (create, reconfigure, execute) ---> GitlabInstance (files, commits, jobs etc)
         |                                                         |
         |----> Files (create, update, delete) <-------------------|
```

The Files code can be in `preview/utils/`

prasadtalasila commented on client/src/preview/store/file.slice.ts:
```diff
@@ -12,28 +14,132 @@ const filesSlice = createSlice({
   initialState,
   reducers: {
-    addOrUpdateFile: (state, action: PayloadAction<FileState>) => {
+    addNewFile: (
```
Please remove duplication and restructure these three reducers.

VanessaScherma replied:
Done. Now there is only the `addOrUpdateFile` reducer, capable to handle these three cases.

prasadtalasila commented:
The correct dependency graph is:

```txt
DigitalTwin (create, reconfigure, execute) ---> GitlabInstance (files, commits, jobs etc)
         |                                                        ^
         |                                                        |
         |----> Files (create, update, delete) -------------------|
```
You can create an interface **IFiles** and then have both **GitlabInstance** and **Files** implement this interface.

prasadtalasila commented:
The current code structure requires addition of child pipeline in parent pipeline. Thus each new DT created must have an entry in top-level [.gitlab-ci.yml](https://maestro.cps.digit.au.dk/gitlab/dtaas/TestUserDTaaS/-/blob/main/.gitlab-ci.yml?ref_type=heads). Otherwise, the execution of new DTs fails.

The solution is to have the create step has update the `.gitlab-ci.yml` at the root-level. Otherwise, the new DTs do not get executed.

VanessaScherma commented on client/src/preview/route/digitaltwins/editor/Sidebar.tsx:
The file was split in this way:
- `Sidebar.tsx`: contains the Sidebar layout and the logic to manage the rendering of files
- `SidebarDialog.tsx`: contains the dialog for adding a new file

VanessaScherma commented on client/src/preview/route/digitaltwins/DigitalTwinsPreview.tsx:
```diff
@@ -1,96 +1,76 @@
-  });
-};
-
 export const DTContent = () => {
```
The purpose of this file has remained unchanged, I have only added the necessary logic for the Create page (the useState `newDigitalTwinName`) while I have moved the pre-existing logic in the function `fetchAssetsAndCreateTwins` to the AssetBoard, as it no longer has to be called up only when the main component (DigitalTwinsPreview) is opened, but each time an AssetBoard is rendered, i.e. when switching to the Manage or Execute tabs. This is necessary in order to allow the new DT to be displayed immediately after creation by switching to one of the aforementioned tabs.

prasadtalasila replied:
The `AssetBoard` is a react component. It is better to move the actual logic out of it.

VanessaScherma commented on client/src/preview/components/asset/AssetBoard.tsx:
```diff
@@ -44,10 +47,43 @@ const AssetGridItem: React.FC<{
+  const [error, setError] = React.useState<string | null>(null);
   const dispatch = useDispatch();
 
+  const gitlabInstance = new GitlabInstance(
```
It has been moved to `init.ts` in `preview/util.`

VanessaScherma commented on client/src/preview/route/digitaltwins/create/ChangeFileNameDialog.tsx:
Functions with file-related operations have been moved to `file.ts` in `preview/util`.

VanessaScherma commented:
> The correct dependency graph is:
> 
> ```
> DigitalTwin (create, reconfigure, execute) ---> GitlabInstance (files, commits, jobs etc)
>          |                                                        ^
>          |                                                        |
>          |----> Files (create, update, delete) -------------------|
> ```
> 
> You can create an interface **IFiles** and then have both **GitlabInstance** and **Files** implement this interface.

This is how the code was refactored:
The methods concerning files, previously placed in the DigitalTwin class, have now been moved to a `FileHandler` class. This class implements the `IFiles` interface. Each DigitalTwin has a FileHandler among its attributes, and initialises it in the constructor by passing it the same GitlabInstance associated with the DigitalTwin, which is used to call the APIs. 

The only reference to files still present in DigitalTwin are the attributes descriptionFiles, configFiles, lifecycleFiles, since they are the file names closely related to the DT.

VanessaScherma commented:
@prasadtalasila 
The handling of the `.gitlab-ci.yml` file is resolved. When a new DT is created, a `trigger_DTName` section is added to the .gitlab-ci.yml file in the parent pipeline. Similarly, this section is removed when deleting the corresponding DT from the Manage tab.

prasadtalasila commented:
@VanessaScherma please rebase the PR

prasadtalasila commented:
@VanessaScherma The addition of new DT adds entries in top-level `.gitlab-ci.yml` but the format is incorrect. As a result, an execution request on new DT triggers the following error.

![image](https://github.com/user-attachments/assets/ab2364ba-6f5c-49aa-b392-c40cf1f0af63)

The matching lines are:

![image](https://github.com/user-attachments/assets/fdd66ed7-36ee-46e4-b21b-aa0215ea1cd5)

The create request is adding extra tabs at the beginning of new lines. Could this be the error?

prasadtalasila commented:
another consequence is that the execution of existing DTs fail as well due to syntax error in top-level `.gitlab-ci.yml`.

prasadtalasila commented on client/src/preview/util/digitalTwin.ts:
```diff
@@ -77,31 +84,6 @@ class DigitalTwin {
-    this.lastExecutionStatus = 'error';
-  }
-
   async triggerPipeline() {
```
this method should be private

prasadtalasila reviewed (commented):
@VanessaScherma Thanks for the updates. Please see the comments.

prasadtalasila commented on client/src/preview/util/init.ts:
```diff
@@ -0,0 +1,50 @@
+    } else {
+      dispatch(setAssets([]));
+    }
+  } catch (err) {
```
No it is alright. Please keep it as is.

VanessaScherma commented on client/src/preview/util/digitalTwin.ts:
`fileUtils.ts` includes the functions regarding files used by some React components, such as `DTContent`, `ChangeFileNameDialog`, etc...
They have been separated to keep only the React components in their respective files. 

`digitalTwinUtils.ts` includes some functions that are not DigitalTwin methods, but are used by the methods themselves. They have been separated to keep the code complexity below the threshold required by CodeClimate.

As a result, I do not believe that unifying them into a `DTFiles`/`DTAssets` class would be an appropriate solution.

prasadtalasila replied:
Ok thanks. Let us discuss this in the coming meeting.

VanessaScherma commented on client/src/preview/util/ifile.ts:
```diff
@@ -0,0 +1,21 @@
+    filePath: string,
+    commitMessage: string,
+  ): Promise<void>;
+  getDescriptionFiles(): Promise<string[]>;
```
`getFileContent(...)` is for getting the contents of a file, so it is not applicable here. I refactored by creating a function `getFileNames(...)`, which receives the file type as a parameter. This returns the file names for the respective category.

prasadtalasila commented:
@VanessaScherma Apparently the tests inside and outside of `test/preview` folder are interacting. I've deleted the `test/preview` directory and ran `yarn test:unit` and `yarn test:int`. Please see the results in the log file.
[test-log.log](https://github.com/user-attachments/files/17719593/test-log.log)
